diff --git a/lib/pool/pool.js b/lib/pool/pool.js
index 659d80681ce11387e3d4eee645c5cbf3df7bcb30..bad49c310096fd02ecaddf907642aec78c44ced8 100644
--- a/lib/pool/pool.js
+++ b/lib/pool/pool.js
@@ -134,6 +134,8 @@ var Pool = /** @class */ (function () {
                     request.reject((0, neo4j_driver_core_1.newError)("Connection acquisition timed out in ".concat(_this._acquisitionTimeout, " ms. Pool status: Active conn count = ").concat(activeCount, ", Idle conn count = ").concat(idleCount, ".")));
                 }
             }, _this._acquisitionTimeout);
+            // https://github.com/neo4j/neo4j-javascript-driver/pull/1196
+            timeoutId.unref();
             request = new PendingRequest(key, acquisitionContext, config, resolve, reject, timeoutId, _this._log);
             allRequests[key].push(request);
             _this._processPendingAcquireRequests(address);
