diff --git a/lib/transaction.js b/lib/transaction.js
index 8c97e77e79bc40856fc51c0f0d315fd6e4e5e763..1d16fcac7009d9c506bcdbabdef7890b2c8296f2 100644
--- a/lib/transaction.js
+++ b/lib/transaction.js
@@ -284,6 +284,12 @@ var Transaction = /** @class */ (function () {
         // error will be "acknowledged" by sending a RESET message
         // database will then forget about this transaction and cleanup all corresponding resources
         // it is thus safe to move this transaction to a FAILED state and disallow any further interactions with it
+
+        if (this._state === _states.FAILED) {
+            // already failed, nothing to do
+            // if we call onError for each result again, we might run into an infinite loop, that causes an OOM eventually
+            return Promise.resolve(null)
+        }
         this._state = _states.FAILED;
         this._onClose();
         this._results.forEach(function (result) {
