diff --git a/dist/schema-builder/storages/type-metadata.storage.js b/dist/schema-builder/storages/type-metadata.storage.js
index 66e83885223152491759b51e8c9118dd76c1a8da..3baefc3bd5c3ca5cb5a833eae4bd0a61db5f1e2c 100644
--- a/dist/schema-builder/storages/type-metadata.storage.js
+++ b/dist/schema-builder/storages/type-metadata.storage.js
@@ -7,6 +7,7 @@ const plugin_constants_1 = require("../../plugin/plugin-constants");
 const collections_1 = require("../collections/");
 const cannot_determine_host_type_error_1 = require("../errors/cannot-determine-host-type.error");
 const undefined_type_error_1 = require("../errors/undefined-type.error");
+const get_interfaces_array_util_1 = require("../utils/get-interfaces-array.util");
 const is_throwing_util_1 = require("../utils/is-throwing.util");
 class TypeMetadataStorageHost {
     constructor() {
@@ -275,6 +276,10 @@ class TypeMetadataStorageHost {
             };
             this.addClassFieldMetadata(fieldMetadata);
             objectOrInterfaceTypeMetadata.properties.push(fieldMetadata);
+            // Add field to other interfaces implementing this interface
+            this.metadataByTargetCollection.all.interface
+                .filter(i => get_interfaces_array_util_1.getInterfacesArray(i.interfaces).includes(target))
+                .forEach(i => i.properties.push(fieldMetadata));
         }
         else {
             const isEmpty = (arr) => arr.length === 0;
