diff --git a/dist/schema-builder/factories/interface-definition.factory.js b/dist/schema-builder/factories/interface-definition.factory.js
index a2dc0aa2498d8980fefc6778ca79bf52d1a6e78e..f1912922b78c71b3dbb68dca8769cee54152cf80 100644
--- a/dist/schema-builder/factories/interface-definition.factory.js
+++ b/dist/schema-builder/factories/interface-definition.factory.js
@@ -82,7 +82,15 @@ let InterfaceDefinitionFactory = class InterfaceDefinitionFactory {
         };
         return () => {
             let fields = {};
-            metadata.properties.forEach((field) => {
+            let properties = [];
+            if (metadata.interfaces) {
+                const implementedInterfaces = type_metadata_storage_1.TypeMetadataStorage.getInterfacesMetadata()
+                    .filter((it) => (0, get_interfaces_array_util_1.getInterfacesArray)(metadata.interfaces).includes(it.target))
+                    .map((it) => it.properties);
+                implementedInterfaces.forEach(fields => properties.push(...(fields || [])));
+            }
+            properties = properties.concat(metadata.properties);
+            properties.forEach((field) => {
                 const type = this.outputTypeFactory.create(field.name, field.typeFn(), options, field.options);
                 fields[field.schemaName] = {
                     description: field.description,
